File: 3_Programação_RPi/3.3_GPIO_Polling/Ex6.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/poll.h>
#include <unistd.h>

int main(void)
{
	struct pollfd pfd[2];
	char buffer;
	int btn_press;
	system("echo 4       > /sys/class/gpio/export");
	system("echo falling > /sys/class/gpio/gpio4/edge");
	system("echo in      > /sys/class/gpio/gpio4/direction");
	system("echo 17      > /sys/class/gpio/export");
	system("echo falling > /sys/class/gpio/gpio17/edge");
	system("echo in      > /sys/class/gpio/gpio17/direction");
	pfd[0].fd = open("/sys/class/gpio/gpio4/value", O_RDONLY);
	if(pfd[0].fd < 0)
	{
		puts("Erro abrindo /sys/class/gpio/gpio4/value");
		puts("Execute este programa como root.");
		return 1;
	}
	pfd[1].fd = open("/sys/class/gpio/gpio17/value", O_RDONLY);
	pfd[0].events = POLLPRI | POLLERR;
	pfd[1].events = POLLPRI | POLLERR;
	puts("Pressione o botao...");
	for(btn_press=0; btn_press<5; btn_press++)
	{
		pfd[0].revents = 0;
		pfd[1].revents = 0;
		lseek(pfd[0].fd, 0, SEEK_SET);
		read(pfd[0].fd, &buffer, 1);
		lseek(pfd[1].fd, 0, SEEK_SET);
		read(pfd[1].fd, &buffer, 1);
		poll(pfd, 2, -1);
		if((pfd[0].revents)&(POLLPRI|POLLERR))
			puts("Borda de descida em GPIO4!");
		if((pfd[1].revents)&(POLLPRI|POLLERR))
			puts("Borda de descida em GPIO17!");
	}
	close(pfd[0].fd);
	close(pfd[1].fd);
	system("echo 4  > /sys/class/gpio/unexport");
	system("echo 17 > /sys/class/gpio/unexport");
	return 0;
}

File: 3_Programação_RPi/3.3_GPIO_Polling/Python/Ex6.py
import RPi.GPIO as gpio
import os, select, time

pins = [4, 17]
os.system("echo %d      > /sys/class/gpio/export" % pins[0])
os.system("echo falling > /sys/class/gpio/gpio%d/edge" % pins[0])
os.system("echo in      > /sys/class/gpio/gpio%d/direction" % pins[0])
os.system("echo %d      > /sys/class/gpio/export" % pins[1])
os.system("echo falling > /sys/class/gpio/gpio%d/edge" % pins[1])
os.system("echo in      > /sys/class/gpio/gpio%d/direction" % pins[1])

fds = [0,0]
fds[0] = os.open("/sys/class/gpio/gpio%d/value" % pins[0], os.O_RDONLY)
fds[1] = os.open("/sys/class/gpio/gpio%d/value" % pins[1], os.O_RDONLY)

pfd = select.poll()
pfd.register(fds[0], select.POLLPRI | select.POLLERR)
pfd.register(fds[1], select.POLLPRI | select.POLLERR)

print("Pressione o botao...")
for btn_press in range(5):
	os.lseek(fds[0], 0, os.SEEK_SET)
	os.read(fds[0], 1)
	os.lseek(fds[1], 0, os.SEEK_SET)
	os.read(fds[1], 1)
	ret = pfd.poll()
	for r in ret:
		if (r[0] == fds[0]) and (r[1] & (select.POLLPRI | select.POLLERR)):
			print("Borda de descida em GPIO%d!" % pins[0])
		if (r[0] == fds[1]) and (r[1] & (select.POLLPRI | select.POLLERR)):
			print("Borda de descida em GPIO%d!" % pins[1])
	time.sleep(0.3)

os.close(fds[0])
os.system("echo %d > /sys/class/gpio/unexport" % pins[0])
os.close(fds[1])
os.system("echo %d > /sys/class/gpio/unexport" % pins[1])
